src/coreclr/jit/optimizer.cpp

```diff
@@ -3870,50 +3905,65 @@ PhaseStatus Compiler::optUnrollLoops()
+            // block will be set to exist within the parent loop, if any. However, if we unroll a loop that has
+            // nested loops, we will create multiple copies of the nested loops. This requires adding new loop table
+            // entries to represent the new loops. Instead of trying to do this incrementally, in the case where
+            // nested loops exist (in any unrolled loop) we rebuild the entire loop table after unrolling.
 
-            for (lval = lbeg; totalIter; totalIter--)
+            BlockToBlockMap        blockMap(getAllocator(CMK_LoopOpt));
+            BasicBlock*            insertAfter                    = bottom;
+            BasicBlock::loopNumber newLoopNum                     = loop.lpParent;
+            bool                   anyNestedLoopsUnrolledThisLoop = false;
+            int                    lval;
+
+            for (lval = lbeg; totalIter > 0; totalIter--)
             {
-                for (block = head->bbNext;; block = block->bbNext)
+                // Note: we can't use the loop.LoopBlocks() iterator, as it captures loop.lpBottom->bbNext at the
+                // beginning of iteration, and we insert blocks before that. So we need to evaluate lpBottom->bbNext
+                // every iteration.
+                for (BasicBlock* block = loop.lpTop; block != loop.lpBottom->bbNext; block = block->bbNext)
                 {
                     BasicBlock* newBlock = insertAfter =
                         fgNewBBafter(block->bbJumpKind, insertAfter, /*extendRegion*/ true);
                     blockMap.Set(block, newBlock, BlockToBlockMap::Overwrite);
 
                     if (!BasicBlock::CloneBlockState(this, newBlock, block, lvar, lval))
                     {
-                        // cloneExpr doesn't handle everything
+                        // CloneBlockState (specifically, gtCloneExpr) doesn't handle everything. If it fails
+                        // to clone a block in the loop, splice out and forget all the blocks we cloned so far:
+                        // put the loop blocks back to how they were before we started cloning blocks,
+                        // and abort unrolling the loop.
                         BasicBlock* oldBottomNext = insertAfter->bbNext;
                         bottom->bbNext            = oldBottomNext;
                         oldBottomNext->bbPrev     = bottom;
-                        optLoopTable[lnum].lpFlags |= LPFLG_DONT_UNROLL;
+                        loop.lpFlags |= LPFLG_DONT_UNROLL; // Mark it so we don't try to unroll it again.
                         INDEBUG(++unrollFailures);
                         JITDUMP("Failed to unroll loop " FMT_LP ": block cloning failed on " FMT_BB "\n", lnum,
                                 block->bbNum);
                         goto DONE_LOOP;
                     }
 
+                    // All blocks in the unrolled loop and all children loops will now be marked
```

Comment:
`children` here confuses me -- are we unrolling a loop that has children...?

Seems like maybe we unroll a loop that has a cloned loop inside where we've unrolled the clone fast path? -- so both the unrolled fast path and the (loop but not marked as loop slow path) get reparented? Still, seems like a lot of code expansion, especially if we create N copies of the slow path.

Reply:
We will unroll a loop that has nested child loops. If there are child loops, I later will force a re-generation of the loop table, so this code to apply new loop number to the unrolled loop block will only be permanent if there are no child loops. The idea here is that if there are no child loops, it is "easy" to set the correct loop number, and there is no need to pay the expense of rebuilding the loop table.

Now, it turns out that we (currently) very rarely unroll loops, but one case where we always unroll is a loop statically determined to execute exactly once. That still goes through all the mechanism, and there might be nested loops within that.

I could expand on the comment here if you thing that would help.

Reply:
A more expansive comment would be helpful.

Is it just the "single iteration" case where we might have nested loops? 

I guess unrolling there makes sense. Seems like for that case -- if we're not already doing so -- we should change the loop exit test to always exit, instead of cloning and then deleting the original loop.

I suppose if unrolling were smart enough to make N-1 copies this would just fall out, and we would not need to zap the originals either -- just repair the loop exit branch as above. 

Again, maybe this already happens?

Reply:
We can have nested loops in any case, not just the single iteration case. Consider:
```
public static int test_01()
{
    int a = 0;
    for (int i = 0; i < 2; i++)
    {
        a += 9;
        for (int j = 0; j < 10000; j++)
        {
            a += 1;
        }
    }
    return a;
}
```
With our current heuristics, we won't unroll this. With `COMPlus_JitStressModeNames=STRESS_UNROLL_LOOPS` we will unroll the outer loop but not the inner. Replace `2` by `0` (zero) and we'll remove the outer loop but also need to remove the inner.

It turns out you can "game" our heuristics as well, e.g.:
```
public static int test_02()
{
    int a = 0;
    for (int i = 0; i < Vector<int>.Count; i++)
    {
        a += 9;
        for (int j = 0; j < 10000; j++)
        {
            a += 1;
        }
    }
    return a;
}
```
We will always unroll a constant loop with `Vector<T>.Count` constant limit, here 8, but we won't unroll the inner loop.

The current code always creates `N` copies of the original code and then eliminates the original loop. It does seem like it could make `N-1` copies instead.